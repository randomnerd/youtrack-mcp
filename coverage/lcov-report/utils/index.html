
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> utils</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">37.04% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>123/332</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">27.81% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>37/133</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">33.33% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>27/81</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">40.06% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>121/302</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="controller-utils.ts"><a href="controller-utils.ts.html">controller-utils.ts</a></td>
	<td data-value="46.66" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 46%"></div><div class="cover-empty" style="width: 54%"></div></div>
	</td>
	<td data-value="46.66" class="pct low">46.66%</td>
	<td data-value="30" class="abs low">14/30</td>
	<td data-value="11.11" class="pct low">11.11%</td>
	<td data-value="18" class="abs low">2/18</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="9" class="abs low">3/9</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="28" class="abs medium">14/28</td>
	</tr>

<tr>
	<td class="file low" data-value="field-builder.ts"><a href="field-builder.ts.html">field-builder.ts</a></td>
	<td data-value="41.9" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 41%"></div><div class="cover-empty" style="width: 59%"></div></div>
	</td>
	<td data-value="41.9" class="pct low">41.9%</td>
	<td data-value="105" class="abs low">44/105</td>
	<td data-value="43.33" class="pct low">43.33%</td>
	<td data-value="30" class="abs low">13/30</td>
	<td data-value="52.94" class="pct medium">52.94%</td>
	<td data-value="17" class="abs medium">9/17</td>
	<td data-value="43.43" class="pct low">43.43%</td>
	<td data-value="99" class="abs low">43/99</td>
	</tr>

<tr>
	<td class="file medium" data-value="view-utils.ts"><a href="view-utils.ts.html">view-utils.ts</a></td>
	<td data-value="62.5" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 62%"></div><div class="cover-empty" style="width: 38%"></div></div>
	</td>
	<td data-value="62.5" class="pct medium">62.5%</td>
	<td data-value="16" class="abs medium">10/16</td>
	<td data-value="55.55" class="pct medium">55.55%</td>
	<td data-value="9" class="abs medium">5/9</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="8" class="abs low">2/8</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="15" class="abs medium">10/15</td>
	</tr>

<tr>
	<td class="file low" data-value="youtrack.ts"><a href="youtrack.ts.html">youtrack.ts</a></td>
	<td data-value="30.38" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 30%"></div><div class="cover-empty" style="width: 70%"></div></div>
	</td>
	<td data-value="30.38" class="pct low">30.38%</td>
	<td data-value="181" class="abs low">55/181</td>
	<td data-value="22.36" class="pct low">22.36%</td>
	<td data-value="76" class="abs low">17/76</td>
	<td data-value="27.65" class="pct low">27.65%</td>
	<td data-value="47" class="abs low">13/47</td>
	<td data-value="33.75" class="pct low">33.75%</td>
	<td data-value="160" class="abs low">54/160</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-05-03T03:08:27.013Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    