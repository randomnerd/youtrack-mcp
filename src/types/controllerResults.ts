import * as YouTrackTypes from './youtrack';

// Pagination interface for list results
export interface Pagination {
  limit?: number;
  skip?: number;
  totalItems?: number;
}

// Common result interface for all controllers
export interface ControllerResult<T> {
  success: boolean;
  data?: T;
  error?: string;
  metadata?: Record<string, unknown>;
}

// Request type for controller handlers
export interface Request {
  params: Record<string, string>;
  query?: Record<string, string | string[]>;
  body?: Record<string, unknown>;
  method?: string;
  headers?: Record<string, string>;
}

// Issue Controller Result Types
export interface IssueDetailResult {
  issue: YouTrackTypes.IssueWithActivities;
  activities?: YouTrackTypes.Activity[];
}

export interface IssueListResult {
  issues: YouTrackTypes.Issue[];
  total: number;
  query?: string;
  title?: string;
  pagination?: Pagination;
}

export interface IssueUpdateResult {
  issueId: string;
  updated: boolean;
}

// Project Controller Result Types
export interface ProjectDetailResult {
  project: YouTrackTypes.Project;
}

export interface ProjectListResult {
  projects: YouTrackTypes.Project[];
  total: number;
  pagination?: Pagination;
}

// Sprint Controller Result Types
export interface SprintDetailResult {
  sprint: YouTrackTypes.Sprint;
  boardId: string;
}

export interface SprintListResult {
  sprints: YouTrackTypes.Sprint[];
  total: number;
  pagination?: Pagination;
  boardId?: string;
  boardName?: string;
}

// Board Controller Result Types
export interface BoardDetailResult {
  board: YouTrackTypes.Board;
  columns?: YouTrackTypes.AgileColumn[];
  issues?: YouTrackTypes.Issue[];
}

export interface BoardListResult {
  boards: YouTrackTypes.Board[];
  total: number;
  pagination?: Pagination;
} 