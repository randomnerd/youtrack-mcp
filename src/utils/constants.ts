/**
 * Pagination constants
 */
export const PAGINATION_LIMITS = {
  ISSUES: 50,
  PROJECTS: 100,
  BOARDS: 50,
  SPRINTS: 50,
  COMMENTS: 100,
  ATTACHMENTS: 100,
  ACTIVITIES: 100,
  DEFAULT: 50
};

/**
 * Default pagination values
 */
export const DEFAULT_PAGINATION = {
  LIMIT: 10,
  SKIP: 0
}; 