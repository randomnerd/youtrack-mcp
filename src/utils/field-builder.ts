/**
 * Utility functions for parsing and serializing YouTrack field strings
 */
export interface FieldNode {
  name: string;
  children?: FieldNode[];
}

/**
 * Builder class for creating field strings
 */
export class FieldBuilder {
  public rootFields: FieldNode[] = [];

  /**
   * Create a new FieldBuilder
   * @param fieldString - Optional serialized field string to initialize with
   */
  constructor(fieldString?: string) {
    if (fieldString) {
      this.rootFields = FieldBuilder.parseFieldString(fieldString);
    }
  }

  /**
   * Parse a YouTrack field string into a structured object
   * @param fieldString - The field string to parse
   * @returns The parsed field structure
   */
  public static parseFieldString(fieldString: string): FieldNode[] {
    let current = 0;
    
    function parseFields(): FieldNode[] {
      const fields: FieldNode[] = [];
      let fieldName = '';
      
      while (current < fieldString.length) {
        const char = fieldString[current];
        
        if (char === '(') {
          // Start of children
          current++; // Skip the opening parenthesis
          const children = parseFields();
          fields.push({ name: fieldName.trim(), children });
          fieldName = '';
        } else if (char === ')') {
          // End of current level
          current++; // Skip the closing parenthesis
          return fields;
        } else if (char === ',') {
          // Field separator
          if (fieldName.trim()) {
            fields.push({ name: fieldName.trim() });
          }
          fieldName = '';
          current++;
        } else {
          // Part of field name
          fieldName += char;
          current++;
        }
      }
      
      // Add the last field if any
      if (fieldName.trim()) {
        fields.push({ name: fieldName.trim() });
      }
      
      return fields;
    }
    
    return parseFields();
  }

  /**
   * Serialize a field structure back to a YouTrack field string
   * @param fields - The field structure to serialize
   * @returns The serialized field string
   */
  public static serializeFieldStructure(fields: FieldNode[]): string {
    return fields.map(field => {
      if (!field.children || field.children.length === 0) {
        return field.name;
      }
      return `${field.name}(${FieldBuilder.serializeFieldStructure(field.children)})`;
    }).join(',');
  }

  /**
   * Add a field to an existing field string
   * @param fieldString - The original field string
   * @param path - Path to the parent field, separated by dots (e.g., "project.fields")
   * @param newField - The new field to add
   * @returns The updated field string
   */
  public static addField(fieldString: string, path: string, newField: string): string {
    const fields = FieldBuilder.parseFieldString(fieldString);
    
    function findAndAddField(nodes: FieldNode[], pathParts: string[]): boolean {
      if (pathParts.length === 0) {
        return false;
      }
      
      const currentPath = pathParts[0];
      const remainingPath = pathParts.slice(1);
      
      for (const node of nodes) {
        if (node.name === currentPath) {
          if (remainingPath.length === 0) {
            // We've found the target node, add the new field
            const newFieldParts = FieldBuilder.parseFieldString(newField);
            if (!node.children) {
              node.children = [];
            }
            node.children.push(...newFieldParts);
            return true;
          } else {
            // Continue down the path
            if (!node.children) {
              node.children = [];
            }
            return findAndAddField(node.children, remainingPath);
          }
        }
      }
      
      return false;
    }
    
    const pathParts = path.split('.');
    findAndAddField(fields, pathParts);
    
    return FieldBuilder.serializeFieldStructure(fields);
  }

  /**
   * Remove a field from an existing field string
   * @param fieldString - The original field string
   * @param fieldPath - Full path to the field to remove, separated by dots
   * @returns The updated field string
   */
  public static removeField(fieldString: string, fieldPath: string): string {
    const fields = FieldBuilder.parseFieldString(fieldString);
    const pathParts = fieldPath.split('.');
    
    function findAndRemoveField(nodes: FieldNode[], parts: string[], index: number): boolean {
      if (index >= parts.length) return false;
      
      const currentPart = parts[index];
      const isLastPart = index === parts.length - 1;
      
      for (let i = 0; i < nodes.length; i++) {
        if (nodes[i].name === currentPart) {
          if (isLastPart) {
            // Found the field to remove
            nodes.splice(i, 1);
            return true;
          } else {
            // Continue searching in children
            if (!nodes[i].children) {
              return false;
            }
            return findAndRemoveField(nodes[i].children!, parts, index + 1);
          }
        }
      }
      
      return false;
    }
    
    findAndRemoveField(fields, pathParts, 0);
    return FieldBuilder.serializeFieldStructure(fields);
  }

  /**
   * Add a field to the builder
   * @param field - Field name or path (dot notation for nested fields)
   * @returns The builder instance for chaining
   */
  public add(field: string): FieldBuilder {
    if (!field) return this;

    // If the field contains dots, it's a nested field
    if (field.includes('.')) {
      this.addNestedField(field);
    } else {
      // Check if this field already exists at the root level
      const exists = this.rootFields.some(node => node.name === field);
      
      // Only add if it doesn't already exist
      if (!exists) {
        // Simple field without children
        this.rootFields.push({ name: field });
      }
    }
    return this;
  }

  /**
   * Add multiple fields at once
   * @param fields - Array of field names or paths
   * @returns The builder instance for chaining
   */
  public addFields(fields: string[]): FieldBuilder {
    fields.forEach(field => this.add(field));
    return this;
  }

  /**
   * Parse an existing field string and add it to the builder
   * @param fieldString - The field string to parse and add
   * @returns The builder instance for chaining
   */
  public addFieldString(fieldString: string): FieldBuilder {
    const parsedFields = FieldBuilder.parseFieldString(fieldString);
    
    // Add each field individually to ensure deduplication
    parsedFields.forEach(field => {
      const exists = this.rootFields.some(node => node.name === field.name);
      if (!exists) {
        this.rootFields.push(field);
      }
    });
    
    return this;
  }

  /**
   * Build the final field string
   * @returns The constructed field string
   */
  public build(): string {
    return FieldBuilder.serializeFieldStructure(this.rootFields);
  }

  /**
   * Add a nested field using dot notation
   * @param fieldPath - The field path (e.g., "project.name.fullName")
   */
  private addNestedField(fieldPath: string): void {
    const parts = fieldPath.split('.');
    let currentLevel = this.rootFields;
    let currentPath = '';

    for (let i = 0; i < parts.length; i++) {
      const part = parts[i];
      currentPath = currentPath ? `${currentPath}.${part}` : part;
      
      // Check if this part already exists at the current level
      let existingNode = currentLevel.find(node => node.name === part);
      
      if (!existingNode) {
        // Create new node if it doesn't exist
        existingNode = { name: part, children: [] };
        currentLevel.push(existingNode);
      }
      
      // Move to the next level
      if (!existingNode.children) {
        existingNode.children = [];
      }
      currentLevel = existingNode.children;
    }
  }

  /**
   * Create a new builder from an existing field string
   * @param fieldString - The existing field string
   * @returns A new FieldBuilder with the parsed fields
   */
  public static from(fieldString: string): FieldBuilder {
    const builder = new FieldBuilder();
    builder.rootFields = FieldBuilder.parseFieldString(fieldString);
    return builder;
  }
} 

// Export helper functions as standalone for backward compatibility
export const parseFieldString = FieldBuilder.parseFieldString;
export const serializeFieldStructure = FieldBuilder.serializeFieldStructure;
export const addField = FieldBuilder.addField;
export const removeField = FieldBuilder.removeField; 