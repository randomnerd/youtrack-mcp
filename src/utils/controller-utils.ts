import { McpResponse, ResourceResponse } from '../views/common';
import { ControllerResult } from '../types/controllerResults';

/**
 * Safely extracts a parameter value from request params
 * @param params Request parameters
 * @param paramName Name of the parameter to extract
 * @returns The parameter value or undefined
 */
export function extractParam(params: Record<string, string | string[] | undefined>, paramName: string): string | undefined {
  if (!params) return undefined;
  
  const param = params[paramName];
  if (typeof param === 'string') {
    return param;
  } else if (Array.isArray(param)) {
    return param[0];
  }
  return undefined;
}

/**
 * Creates a standard error response for controllers
 * @param error Error object or string
 * @param messagePrefix Optional prefix for the error message
 * @returns McpResponse with error information
 */
export function createErrorResponse(error: unknown, messagePrefix?: string): McpResponse {
  const errorMessage = error instanceof Error ? error.message : String(error || 'Unknown error');
  const prefix = messagePrefix ? `${messagePrefix}: ` : '';
  
  return {
    content: [{ type: 'text', text: `${prefix}${errorMessage}` }],
    isError: true
  };
}

/**
 * Creates a controller error result
 * @param error Error object or string
 * @param messagePrefix Optional prefix for the error message
 * @returns ControllerResult with error information
 */
export function createErrorResult<T>(error: unknown, messagePrefix?: string): ControllerResult<T> {
  const errorMessage = error instanceof Error ? error.message : String(error || 'Unknown error');
  const prefix = messagePrefix ? `${messagePrefix}: ` : '';
  
  return {
    success: false,
    error: `${prefix}${errorMessage}`
  };
}

/**
 * Creates a standard resource error response
 * @param uri The resource URI
 * @param error Error object or string
 * @returns ResourceResponse with error information
 */
export function createResourceErrorResponse(uri: URL, error: unknown): ResourceResponse {
  const errorMessage = error instanceof Error ? error.message : String(error || 'Unknown error');
  
  return {
    contents: [{
      uri: uri.href,
      text: `Error: ${errorMessage}`
    }]
  };
}

/**
 * Creates a standard resource response
 * @param uri The resource URI
 * @param text Response text content
 * @returns ResourceResponse with the provided content
 */
export function createResourceResponse(uri: URL, text: string): ResourceResponse {
  return {
    contents: [{
      uri: uri.href,
      text
    }]
  };
}

/**
 * Wraps controller methods with standard error handling
 * @param fn The async function to wrap
 * @param errorPrefix Optional prefix for error messages
 * @returns A function with standardized error handling
 */
export function withErrorHandling<T extends any[], R>(
  fn: (...args: T) => Promise<ControllerResult<R>>,
  errorPrefix?: string
): (...args: T) => Promise<ControllerResult<R>> {
  return async (...args: T): Promise<ControllerResult<R>> => {
    try {
      return await fn(...args);
    } catch (error) {
      return createErrorResult<R>(error, errorPrefix);
    }
  };
}

/**
 * Wraps resource handler methods with standard error handling
 * @param fn The resource handler function to wrap
 * @returns A function with standardized error handling for resources
 */
export function withResourceErrorHandling<T extends any[], R>(
  fn: (...args: T) => Promise<R>,
): (...args: T) => Promise<R> {
  return async (...args: T): Promise<R> => {
    try {
      return await fn(...args);
    } catch (error) {
      const uri = args[0] as URL;
      return createResourceErrorResponse(uri, error) as unknown as R;
    }
  };
}

/**
 * Formats a date consistently throughout the application
 * @param date Date string, number, or Date object
 * @returns Formatted date string or 'N/A' if date is invalid
 */
export function formatDate(date: string | number | Date | null | undefined): string {
  if (date === null || date === undefined) return 'N/A';
  try {
    return new Date(date).toLocaleDateString();
  } catch (e) {
    return 'N/A';
  }
} 