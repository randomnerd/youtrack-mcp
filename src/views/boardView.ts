import * as YouTrackTypes from '../types/youtrack';
import { CommonView, McpResponse, ResourceResponse } from './common';
import { formatBoardListItem, formatBoardProjects, formatSprintDetailItem } from '../utils/view-utils';
import { formatYouTrackData } from '../utils/youtrack-json-formatter';
import type { BoardDetailResult, BoardListResult, ControllerResult } from '../types/controllerResults';
export class BoardView {
  static renderList(result: ControllerResult<BoardListResult>): McpResponse {
    if (!result.success || !result.data) {
      return this.renderError(result.error || 'Failed to fetch boards');
    }
    
    const { boards } = result.data;
    
    return {
      content: [
        {
          type: 'text',
          text: formatYouTrackData(boards, { stringify: true }),
        }
      ],
    };
  }

  static renderDetail(result: ControllerResult<BoardDetailResult>): McpResponse {
    if (!result.success || !result.data) {
      return this.renderError(result.error || 'Failed to fetch board');
    }
    
    const { board } = result.data;
    
    return {
      content: [
        {
          type: 'text',
          text: formatYouTrackData(board, { stringify: true }),
        }
      ],
    };
  }

  static renderEmpty = CommonView.renderEmpty;
  static renderError = CommonView.renderError;

  static handleResourceRequest(uri: URL, board?: YouTrackTypes.Board | YouTrackTypes.Board[]): ResourceResponse {
    if (!board) {
      return CommonView.createResourceResponse(uri, "No board data available");
    }

    return CommonView.createResourceResponse(uri, formatYouTrackData(board, { stringify: true }));
  }
} 