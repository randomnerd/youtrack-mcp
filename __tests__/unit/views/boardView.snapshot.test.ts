import { BoardView } from '../../../src/views/boardView';
import { formatYouTrackData } from '../../../src/utils/youtrack-json-formatter';
import { ControllerResult } from '../../../src/types/controllerResults';
import * as YouTrackTypes from '../../../src/types/youtrack';
import { URL } from 'url';

// Use real formatYouTrackData behavior for snapshot tests
jest.mock('../../../src/utils/youtrack-json-formatter', () => ({
  formatYouTrackData: jest.fn().mockImplementation((data, options) => {
    if (options?.stringify) {
      return JSON.stringify(data, null, 2);
    }
    return data;
  })
}));

describe('BoardView Snapshot Tests', () => {
  // Fixed timestamps for consistent snapshots
  const START_TIME_1 = 1746318950222;
  const END_TIME_1 = 1746319450222;
  const START_TIME_2 = 1746319950222;
  const END_TIME_2 = 1746320950222;
  
  const mockBoard: YouTrackTypes.Board = {
    id: 'board-1',
    name: 'Development Board',
    description: 'Main development board for the project',
    $type: 'Agile',
    owner: {
      id: 'user-1',
      name: 'Jane Doe',
      $type: 'User'
    },
    projects: [
      { id: 'project-1', name: 'Frontend', $type: 'Project' },
      { id: 'project-2', name: 'Backend', $type: 'Project' }
    ],
    sprints: [
      {
        id: 'sprint-1',
        name: 'Sprint 1',
        goal: 'Implement core features',
        start: START_TIME_1,
        finish: END_TIME_1,
        $type: 'Sprint'
      },
      {
        id: 'sprint-2',
        name: 'Sprint 2',
        goal: 'Polish UI components',
        start: START_TIME_2,
        finish: END_TIME_2,
        $type: 'Sprint'
      }
    ]
  };

  const mockBoards: YouTrackTypes.Board[] = [
    mockBoard,
    {
      id: 'board-2',
      name: 'Bug Tracking Board',
      description: 'Board for tracking bug fixes',
      $type: 'Agile',
      owner: {
        id: 'user-2',
        name: 'John Smith',
        $type: 'User'
      },
      projects: [
        { id: 'project-1', name: 'Frontend', $type: 'Project' }
      ],
      sprints: []
    }
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('renderList', () => {
    it('should match snapshot for board list', () => {
      const controllerResult = {
        success: true,
        data: {
          boards: mockBoards,
          total: mockBoards.length
        }
      } as ControllerResult<any>;
      
      const result = BoardView.renderList(controllerResult);
      expect(result).toMatchSnapshot();
    });

    it('should match snapshot for empty board list', () => {
      const controllerResult = {
        success: true,
        data: {
          boards: [],
          total: 0
        }
      } as ControllerResult<any>;
      
      const result = BoardView.renderList(controllerResult);
      expect(result).toMatchSnapshot();
    });

    it('should match snapshot for error response', () => {
      const controllerResult = {
        success: false,
        error: 'Failed to fetch boards due to authentication error'
      } as ControllerResult<any>;
      
      const result = BoardView.renderList(controllerResult);
      expect(result).toMatchSnapshot();
    });
  });

  describe('renderDetail', () => {
    it('should match snapshot for board detail', () => {
      const controllerResult = {
        success: true,
        data: {
          board: mockBoard
        }
      } as ControllerResult<any>;
      
      const result = BoardView.renderDetail(controllerResult);
      expect(result).toMatchSnapshot();
    });

    it('should match snapshot for error in board detail', () => {
      const controllerResult = {
        success: false,
        error: 'Board not found'
      } as ControllerResult<any>;
      
      const result = BoardView.renderDetail(controllerResult);
      expect(result).toMatchSnapshot();
    });
  });

  describe('handleResourceRequest', () => {
    const uri = new URL('http://example.com/boards');
    
    it('should match snapshot for board list resource', () => {
      const result = BoardView.handleResourceRequest(uri, mockBoards);
      expect(result).toMatchSnapshot();
    });
    
    it('should match snapshot for single board resource', () => {
      const result = BoardView.handleResourceRequest(uri, mockBoard);
      expect(result).toMatchSnapshot();
    });

    it('should match snapshot for missing board data', () => {
      const result = BoardView.handleResourceRequest(uri, undefined);
      expect(result).toMatchSnapshot();
    });
  });
}); 