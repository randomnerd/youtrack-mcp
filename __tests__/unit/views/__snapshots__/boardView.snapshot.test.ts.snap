// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardView Snapshot Tests handleResourceRequest should match snapshot for board list resource 1`] = `
{
  "contents": [
    {
      "text": "[
  {
    "id": "board-1",
    "name": "Development Board",
    "description": "Main development board for the project",
    "$type": "Agile",
    "owner": {
      "id": "user-1",
      "name": "Jane Doe",
      "$type": "User"
    },
    "projects": [
      {
        "id": "project-1",
        "name": "Frontend",
        "$type": "Project"
      },
      {
        "id": "project-2",
        "name": "Backend",
        "$type": "Project"
      }
    ],
    "sprints": [
      {
        "id": "sprint-1",
        "name": "Sprint 1",
        "goal": "Implement core features",
        "start": 1746318950222,
        "finish": 1746319450222,
        "$type": "Sprint"
      },
      {
        "id": "sprint-2",
        "name": "Sprint 2",
        "goal": "Polish UI components",
        "start": 1746319950222,
        "finish": 1746320950222,
        "$type": "Sprint"
      }
    ]
  },
  {
    "id": "board-2",
    "name": "Bug Tracking Board",
    "description": "Board for tracking bug fixes",
    "$type": "Agile",
    "owner": {
      "id": "user-2",
      "name": "John Smith",
      "$type": "User"
    },
    "projects": [
      {
        "id": "project-1",
        "name": "Frontend",
        "$type": "Project"
      }
    ],
    "sprints": []
  }
]",
      "uri": "http://example.com/boards",
    },
  ],
}
`;

exports[`BoardView Snapshot Tests handleResourceRequest should match snapshot for missing board data 1`] = `
{
  "contents": [
    {
      "text": "No board data available",
      "uri": "http://example.com/boards",
    },
  ],
}
`;

exports[`BoardView Snapshot Tests handleResourceRequest should match snapshot for single board resource 1`] = `
{
  "contents": [
    {
      "text": "{
  "id": "board-1",
  "name": "Development Board",
  "description": "Main development board for the project",
  "$type": "Agile",
  "owner": {
    "id": "user-1",
    "name": "Jane Doe",
    "$type": "User"
  },
  "projects": [
    {
      "id": "project-1",
      "name": "Frontend",
      "$type": "Project"
    },
    {
      "id": "project-2",
      "name": "Backend",
      "$type": "Project"
    }
  ],
  "sprints": [
    {
      "id": "sprint-1",
      "name": "Sprint 1",
      "goal": "Implement core features",
      "start": 1746318950222,
      "finish": 1746319450222,
      "$type": "Sprint"
    },
    {
      "id": "sprint-2",
      "name": "Sprint 2",
      "goal": "Polish UI components",
      "start": 1746319950222,
      "finish": 1746320950222,
      "$type": "Sprint"
    }
  ]
}",
      "uri": "http://example.com/boards",
    },
  ],
}
`;

exports[`BoardView Snapshot Tests renderDetail should match snapshot for board detail 1`] = `
{
  "content": [
    {
      "text": "{
  "id": "board-1",
  "name": "Development Board",
  "description": "Main development board for the project",
  "$type": "Agile",
  "owner": {
    "id": "user-1",
    "name": "Jane Doe",
    "$type": "User"
  },
  "projects": [
    {
      "id": "project-1",
      "name": "Frontend",
      "$type": "Project"
    },
    {
      "id": "project-2",
      "name": "Backend",
      "$type": "Project"
    }
  ],
  "sprints": [
    {
      "id": "sprint-1",
      "name": "Sprint 1",
      "goal": "Implement core features",
      "start": 1746318950222,
      "finish": 1746319450222,
      "$type": "Sprint"
    },
    {
      "id": "sprint-2",
      "name": "Sprint 2",
      "goal": "Polish UI components",
      "start": 1746319950222,
      "finish": 1746320950222,
      "$type": "Sprint"
    }
  ]
}",
      "type": "text",
    },
  ],
}
`;

exports[`BoardView Snapshot Tests renderDetail should match snapshot for error in board detail 1`] = `
{
  "content": [
    {
      "text": "Error: Board not found",
      "type": "text",
    },
  ],
  "isError": true,
}
`;

exports[`BoardView Snapshot Tests renderList should match snapshot for board list 1`] = `
{
  "content": [
    {
      "text": "[
  {
    "id": "board-1",
    "name": "Development Board",
    "description": "Main development board for the project",
    "$type": "Agile",
    "owner": {
      "id": "user-1",
      "name": "Jane Doe",
      "$type": "User"
    },
    "projects": [
      {
        "id": "project-1",
        "name": "Frontend",
        "$type": "Project"
      },
      {
        "id": "project-2",
        "name": "Backend",
        "$type": "Project"
      }
    ],
    "sprints": [
      {
        "id": "sprint-1",
        "name": "Sprint 1",
        "goal": "Implement core features",
        "start": 1746318950222,
        "finish": 1746319450222,
        "$type": "Sprint"
      },
      {
        "id": "sprint-2",
        "name": "Sprint 2",
        "goal": "Polish UI components",
        "start": 1746319950222,
        "finish": 1746320950222,
        "$type": "Sprint"
      }
    ]
  },
  {
    "id": "board-2",
    "name": "Bug Tracking Board",
    "description": "Board for tracking bug fixes",
    "$type": "Agile",
    "owner": {
      "id": "user-2",
      "name": "John Smith",
      "$type": "User"
    },
    "projects": [
      {
        "id": "project-1",
        "name": "Frontend",
        "$type": "Project"
      }
    ],
    "sprints": []
  }
]",
      "type": "text",
    },
  ],
}
`;

exports[`BoardView Snapshot Tests renderList should match snapshot for empty board list 1`] = `
{
  "content": [
    {
      "text": "[]",
      "type": "text",
    },
  ],
}
`;

exports[`BoardView Snapshot Tests renderList should match snapshot for error response 1`] = `
{
  "content": [
    {
      "text": "Error: Failed to fetch boards due to authentication error",
      "type": "text",
    },
  ],
  "isError": true,
}
`;
